File: typescript/robot-name/robot-name.ts
const ALPHA = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'

class RobotFactory {
  names: string[] = [];
  currentNameIndex = 0;

  constructor() {
    this.populateNames()
  }

  shuffleNames = (): void => {
    for (let i = 0; i < this.names.length - 1; i++) {
      const randomIndex = Math.floor(Math.random() * (this.names.length - i)) + i
      let temp = this.names[randomIndex]
      this.names[randomIndex] = this.names[i]
      this.names[i] = temp
    }
  }

  populateNames = (): void => {
    for (let i = 0; i < 26; i++) {
      for (let j = 0; j < 26; j++) {
        for (let x = 0; x < 10; x++) {
          for (let y = 0; y < 10; y++) {
            for (let z = 0; z < 10; z++) {
              this.names.push(`${ALPHA[i]}${ALPHA[j]}${x}${y}${z}`)
            }
          } 
        }
      }
    }
    this.shuffleNames()
  }

  getName = (): string => this.names[this.currentNameIndex++]
}

const FACTORY = new RobotFactory();

class RobotName {
  name: string
  constructor() {
    this.name = this.getName()
  }

  resetName = () => {
    this.name = this.getName()
  }

  getName = () => FACTORY.getName()
}

export default RobotName

File: kotlin/robot-name/src/main/kotlin/RobotName.kt
class RobotFactory {
    var names = mutableListOf<String>()
    public var index = 0
    init {
        ('A'..'Z').toList().forEach { a ->
            ('A'..'Z').toList().forEach { b ->
                (0..9).toList().forEach { x ->
                    (0..9).toList().forEach { y ->
                        (0..9).toList().forEach { z ->
                            names.add("$a$b$x$y$z")
                        }
                    }
                }
            }
        }
        // randomize robot names with Fisher-Yates shuffle
        (0 until names.size).forEach { it ->
            // find random index from remaining names
            val randomIndex = (it until names.size).random()
            // swap current name with name at random index
            val temp = names[it]
            names[it] = names[randomIndex]
            names[randomIndex] = temp
         }
    }
}

class Robot {
    companion object {
        val robots = RobotFactory()
    }

    var myIndex: Int

    init {
        myIndex = robots.index++
    }

    val name: String
        get() = robots.names[myIndex]

    fun reset() {
        myIndex = robots.index++
    }
}
